obsolete/weather/weather/wm918/wx200d-1.2/wx200.1
.\" @(#)$Id: wx200.1,v 1.1 2002/09/23 19:12:51 bogdan Rel $
.\"
.\" Copyright (C) 1998 - 2001 Tim Witham <twitham@quiknet.com>
.\"
.\" (see the files README and COPYING for more details)

.TH WX200 1 "Sep 05, 2001" "wx200d Manual"
.SH NAME
wx200 \- wx200d weather client
.SH SYNOPSIS
.B wx200
[-h | --help] [-r | --refresh]
[--dm | --md] [--12hr | --24hr]
[--C | --celsius | --F | --fahrenheit]
[--inhg | --mmhg | --mbar | --hpa]
[--mm | --mm/h | --in | --in/h]
[--mph | --knots | --mps | --kph]
[-e | --noerrors] [-u | --nounits]
[ -a | --alarms | -t | --tab | -l | --line | --aprs |
[[--intemp] [--inhumidity] [--indewpoint]
[--temp] [--humidity] [--dewpoint]
[--local] [--sea]
[--gust] [--average]
[--rate] [--yesterday] [--total]
[--trend] [--forecast]
[--power] [--battery] [--display]]]
[[host][:port]]

.SH DESCRIPTION

.B wx200
is a
.B wx200d(8)
client that displays all current data of the WX200 weather station on
the given host and port.  If host or port is not specified, the
.B WX200D
environment variable is checked for the missing field(s).  Finally, if
neither of these provide the host or port, the built-in defaults
(localhost:9753) are used.
.P

With no options,
.B wx200
displays all current weather data and exits.  The data is formatted as
human readable text 80 columns wide to fit in most standard terminals.
.P

.B wx200
can also output HTML as a CGI program.  If the environment variable
.I QUERY_STRING
is set,
.B wx200
assumes it was called by a WWW server and outputs text/html instead of
plain text.
.P

The display should be mostly self-explanatory.  Sensor values are
followed by a unit designator.  Between the number and the unit is a
space character or one of the following special flag characters:

.TP 0.5i
.B *
The value is Out of Range (this is shown as O.R. on the physical WX200)

.TP 0.5i
.B !
There is a sensor error so the value has not been updated (this is
shown as ---- on the physical WX200)

.TP 0.5i
.B #
Both the * and ! flags are set.  That is, the sensor is out of range
and there is an error.

.SH OPTIONS

.TP 0.5i
.B "-h, --help"
Display a help usage message showing these options, then exit.

.TP 0.5i
.B "-r, --refresh"
Continuously refresh the display of weather data, instead of exiting.
This looks best in an 80x25 terminal with
.B -a
or an 80x14 otherwise.  Under X, an alias like:
.B xterm -name WX200 -geometry 80x14 +sb -e wx200 -r
works great.  If
.B QUERY_STRING
is set, this option adds a 10 second
.I Refresh:
header which causes most web browsers to request the page every 10
seconds.

.TP 0.5i
.B "--dm, --md"
Display dates in Day/Month or Month/Day format respectively.  If
neither of these options are specified, the format of the physical
WX200 is used.

.TP 0.5i
.B "--12hr, --24hr"
Display times in 12 hour (with am/pm indicators) or 24 hour notation
respectively.  If neither of these options are specified, the format
of the physical WX200 is used.

.TP 0.5i
.B "--C, --celsius, --F, --fahrenheit"
Display temperatures in degrees Celsius or degrees Fahrenheit
respectively.  If none of these options are specified, the format of
the physical WX200 is used.

.TP 0.5i
.B "--inhg, --mmhg, --mbar, --hpa"
Display barometric pressure in inches of mercury, millimeters of
mercury, millibars or hecto-pascals respectively.  If none of these
options are specified, the format of the physical WX200 is used.

.TP 0.5i
.B "--mm, --mm/h, --in, --in/h"
Display rain fall depths and rates in inches and inches per hour or
millimeters and millimeters per hour respectively.  If none of these
options are specified, the format of the physical WX200 is used.

.TP 0.5i
.B "--mph, --knots, --mps, --kph"
Display wind speed in miles per hour, nautical miles per hour, meters
per second or kilometers per hour respectively.  If none of these
options are specified, the format of the physical WX200 is used.

.TP 0.5i
.B "-e, --noerrors"
Suppress error and out of range flags from the display.

.TP 0.5i
.B "-u, --nounits"
Suppress units from the display.

.TP 0.5i
.B "-a, --alarms"
Add alarm and flag legend details to the display.

.TP 0.5i
.B "-t, --tab"
Instead of the default display, show all variables on one
tab-delimited line in
.B wx200d(5)
tab format.

.TP 0.5i
.B "--aprs"
Instead of the default display, show one line in amateur radio APRS
network format.  There is an example wxaprs script in the distribution
for transmitting data in this format.

.TP 0.5i
.B "-l, --line"
Instead of the default display, show all weather variables on one
tab-delimited line in the order listed here:

.TP 0.5i
.B "--intemp, --temp, --inhumidity, --humidity, --indewpoint, --dewpoint, --local, --sea, --gust, --average, --rate, --yesterday, --total, --trend, --forecast, --power, --battery, --display"
Instead of the default display, list only the given variables on one
line.  The order of the output is fixed like this and the order of the
options on the command line will make no difference.


.SH EXAMPLES

.TP 0.5i
.B wx200 -a -r

Show all weather station information continuously.

.TP 0.5i
.B wx200 -eu --temp --humidity --dewpoint --C

Output only 3 numbers (no error flags or units) with temperatures in
degress Celsius.

.SH ENVIRONMENT

.TP 0.5i
.B WX200D
If set, the [host][:port] to contact if none is specified on the
command line.

.TP 0.5i
.B QUERY_STRING
If set, causes
.B wx200
to output text/html as a CGI program.
.B QUERY_STRING
can contain any of the above options separated by & or = characters,
as a front-end HTML form may provide.

.SH LIMITATIONS

.SH BUGS

.SH SEE ALSO
wxplot(1), wx200d(8), wx200d(5)

.SH AUTHOR
Tim Witham <twitham@quiknet.com>
.SH COPYRIGHT
Copyright (C) 1998 - 2001 Tim Witham <twitham@quiknet.com>

wx200 is released under the conditions of the GNU General Public
License.  See the files README and COPYING in the distribution for
details.


sip_router/NEWS
XXX content of this file reflects a future release;
    features described here may or may not be available
    in the CVS tree
    

Release notes for SIP Express Router (ser)
***********************************************

$Id: NEWS,v 1.12 2003/04/18 14:11:30 jiri Exp $

***********************************************
* Changes introduced in 0.8.11
***********************************************

+--------------------------------------------------------+
| CAUTION: the 0.8.11 release include changes which      |
| are incompatible with scripts and databases used       |
| in previous versions. Care is advised when upgrading   |
| from previous releases to 0.8.11.                      |
+--------------------------------------------------------+

New features
=============
- RFC3261 support
	- TCP support and cross-transport forwarding [core]
	- loose routing support [rr module]
- New modules
	- vm -- voicemail interface [vm]
	- ENUM support [enum]
	- presence agent [pa]
	- dynamic domain management -- allows to manipulate 
  	  hosting of multiple domains in run-time [module]
	- flat-text-file database support [dbtext]
	- rich access control lists [permissions]
- Feature Improvements
	- click-to-dial, which is based on improved tm/FIFO 
	  that better supports external applications [tm module]
	- web accounting -- acc module can report to serweb
  	  on placed calls [acc module]
	- improved exec module (header fields passed now
      as environment variables to scripts) [exec module]
- Architectural Improvements
	- powerpc fast locking support
	- netbsd support
	- 64 bits arch. support (e.g. netbsd/sparc64).
- New Experimental Features (not tested at all yet)
	- nathelper utility for Cisco/ATA NAT traversal [nathelper]
	- another NAT traversal utility [mangler]
	- postgress support [postgress]
	- fcp module [fcp]
	- pdt module (prefix2domain) [pdt]

Changes to use of ser scripts
=============================

About Multiple Transport Support
--------------------------------
SER now suports multiple transport protocols: UDP and TCP. As there
may be UAs which support only either protocol and cannot speak to
each other directly, we recommend to alway record-route SIP requests,
to keep the transport-translating SER in path. Also, if a destination
transport is not known, stateful forwarding is recommended -- use of
stateless forwarding for TCP2UDP would result in loss of reliability.


core
----
- reply_route has been renamed to failure_route -- the old name caused
  too much confusion
- forward_tcp and forward_udp can force SER to forward via specific
  transport protocol

acc module:
-----------
- radius and sql support integrated in this module; you need to
  recompile to enable it
- acc_flag is now called log_flag to better reflect it relates
  to the syslog mode (as opposed to sql/radius); for the same
  reasons, the accounting action is now called "acc_log_request"
- log_fmt allows now to specify what will be printed to syslog

auth module:
------------
- auth module has been split in auth, auth_db, auth_radius, group
  group_radius, uri and uri_radius 
- all the parameters that were part of former auth module are now 
  part of auth_db module
XXX 

im module:
----------
- im is no longer used and has been obsoleted by TM

exec module:
------------
- exec_uri and exec_user have been obsoleted by exec_dset; 
  exec_dset is identical to exec_uri in capabilities; it 
  additionaly passes content of request elements (header 
  fields and URI parts) in environment variables; users of 
  exec_user can use exec_dset now and use the "URI_USER"  
  variable to learn user part of URI
- exec_dset and exec_msg return false, if return value of 
  script does not euqal zero
- exec_dset takes an additional parameter, which enables 
  validation of SIP URIs returned by external application

jabber module:
--------------
- presence support for Jabber users is enabled loading the PA
  module and using subscribe("jabber") for SUBSCRIBE requests to
  jabber user 

msilo module:
-------------
- m_store has now a parameter to set what should be considered
  for storing as destination uri. This enables support for saving
  the messages on negative replies.

radius_acc module:
------------------
- radius_acc module has been removed and radius accounting 
  is now part of acc module

registrar/usrloc modules:
-------------------------
XXX
  
rr module:
----------
- addRecordRoute has been replaced with record_route
- rewriteFromRoute has been replaced with loose_route()
- a new option, "enable_full_lr" can be set to make life
  with misimplemented UAs easier and put LR in from "lr=on"

tm module:
----------
- t_reply_unsafe, used in former versions within reply_routes,
  is deprecated; now t_reply is used from any places in script
- t_on_negative is renamed to t_on_failure -- the old name just
  caused too much confusion
- FIFO t_uac used by some applications (like serweb) has been
  replaced with t_uac_dlg (which allows easier use by dialog-
  oriented applications, like click-to-dial) 
- if you wish to do forward to another destination from 
  failure_route (reply_route formerly), you need to call t_relay
  or t_relay_to explicitely now
- t_relay_to has been replaced with t_relay_to_udp and t_relay_to_tcp



obsolete/weather/weather/wm918/wx200d-1.2/wx200.c
/*
 * @(#)$Id: wx200.c,v 1.1 2002/09/23 19:12:51 bogdan Rel $
 *
 * Copyright (C) 1998 - 2001 Tim Witham <twitham@quiknet.com>
 *
 * (see the files README and COPYING for more details)
 *
 * This file implements a command-line wx200d client
 *
 */

#include <config.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <getopt.h>
#include <string.h>
#include "wx200.h"

#define MAXARGS 20		/* maximum CGI args to parse */
#define TMPLEN 128		/* max length of CGI */

char *progname;
char *query;

void
usage(int ret)
{
  if (query)
      printf("Content-type: text/plain\nStatus: 200\n\n");
  printf("usage: %s [options] [[host][:port]]
  -h	--help			show this help and exit		VERSION: %s
  -r	--refresh		continuously refresh instead of exiting
  --dm --md --12hr --24hr		date order and time format
  --C --celsius --F --fahrenheit	temperature format
  --inhg --mmhg --mbar --hpa		barometric pressure format
  --mm --mm/h --in --in/h		rain rate and accumulation format
  --mph --knots --mps --kph		wind speed format
  -e	--noerrors		suppress error and out of range flags
  -u	--nounits		suppress units from the display
  -a	--alarms		add all alarms and legend to the display  OR:
  -t	--tab			tab-delimited metric line as logged	  OR:
  -l	--line			all variables on one line in custom units OR:
	--aprs			line in APRS (ham radio) format and units OR:
  --intemp --inhumidity --indewpoint	show indoor variable(s)
  --temp --humidity --dewpoint		show outdoor variable(s)
  --local --sea				show barometric pressure(s)
  --gust --average --chill		show wind variable(s)
  --rate --yesterday --total		show rain variable(s)
  --trend --forecast			show barometer trend / 12-24hr forecast
  --power --battery --display		show power, battery, display status
options may be uniquely abbreviated; units default to those of the WX200
", progname, VERSION);
  exit(ret);
}

int
main(int argc, char **argv)
{
  const char *flags = "Hharlteu";
  int opt, num = EOF, socket, done = 0, opt_a = 0, opt_r = 0, opt_t = 0;
  int index, ftemp = -1, fbaro = -1, fdepth = -1, fspeed = -1, ftd = 0;
  int line = 0, i;
  char *hostport = NULL, buf[300], *string, *arg[MAXARGS + 2];
  char *tmp, tmpbuf[TMPLEN];
  char *disp[] = {"Clock", "Temp", "Humidity", "Dew",
		  "Baro", "Wind", "W.Chill", "Rain"};
  char *acdc[] = {"AC", "DC"};
  char *batt[] = {"ok ", "LOW"};
  char format[] = "%11s:"
    "%7s Hi%7s %s %7s Hi%7s %s\n"
    "            "
    "        Lo%7s %s         Lo%7s %s\n";
  char aformat[] = "                    "
    "Al%7s %7s %3s           Al%7s %7s %3s\n\n";
  char *trend[] = {"Rising", "Steady", "", "Falling"};
  char *pred[] = {"Sunny", "Cloudy", "", "Partly Cloudy",
		  "", "", "", "Rain"};
  char *offon[] = {"Off", "On "};
  int current[] = {0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0};

  struct option longopt[] = {
    {"help", 0, 0, 'h'},
    {"alarms", 0, 0, 'a'},
    {"refresh", 0, 0, 'r'},
    {"line", 0, 0, 'l'},
    {"tab", 0, 0, 't'},
    {"noerrors", 0, 0, 'e'},
    {"nounits", 0, 0, 'u'},
    {"celsius", 0, 0, 1}, {"fahrenheit", 0, 0, 2},
    {"C", 0, 0, 1}, {"F", 0, 0, 2},
    {"inhg", 0, 0, 3}, {"mmhg", 0, 0, 4}, {"mbar", 0, 0, 5}, {"hpa", 0, 0, 6},
    {"mm", 0, 0, 7}, {"mm/h", 0, 0, 7}, {"in", 0, 0, 8}, {"in/h", 0, 0, 8},
    {"mph", 0, 0, 9}, {"knots", 0, 0, 10}, {"mps", 0, 0, 11}, {"kph", 0, 0, 12},
    {"dm", 0, 0, 13}, {"md", 0, 0, 14},
    {"12hr", 0, 0, 15}, {"24hr", 0, 0, 16},
    {"aprs", 0, &current[0], 2},

    /* added by trs: 12 Feb 1998, names changed by twitham 16 Apr 1998 */
    {"intemp", 0, &current[0], 1},
    {"temp", 0, &current[1], 1}, /* can now  do multiple ones 9 Jun 2001 */
    {"inhumidity", 0, &current[2], 1},
    {"humidity", 0, &current[3], 1},
    {"indewpoint", 0, &current[4], 1},
    {"dewpoint", 0, &current[5], 1},
    {"local", 0, &current[6], 1},
    {"sea", 0, &current[7], 1},
    {"gust", 0, &current[8], 1},
    {"average", 0, &current[9], 1},
    {"chill", 0, &current[10], 1},
    {"rate", 0, &current[11], 1},
    {"yesterday", 0, &current[12], 1},
    {"total", 0, &current[13], 1},
    {"trend", 0, &current[14], 1},
    {"forecast", 0, &current[15], 1},
    {"power", 0, &current[16], 1},
    {"battery", 0, &current[17], 1},
    {"display", 0, &current[18], 1},
    {0, 0, 0, 0}
  };

  progname = strrchr(argv[0], '/');
  if (progname == NULL)
    progname = argv[0];
  else
    progname++;

  if ((query = getenv("QUERY_STRING"))) { /* we running as a web CGI ? */
    strncpy(tmpbuf, query, TMPLEN);
    tmp = tmpbuf;
    argv = arg;			/* get args from CGI env */
    argv[argc = 0] = progname;
    while ((argc < MAXARGS) && (string = strtok(argc ? NULL : tmp, "&="))) {
      if (string[0] != '\0') {
	argc++;
	argv[argc] = string;
      }
    }
    argv[++argc] = NULL;
  }

  while ((opt = getopt_long(argc, argv, flags, longopt, &index)) != EOF) {
    switch (opt) {		/* parse command-line or CGI options */
    case '?':
      usage(1);
    case 'h':
    case 'H':
      usage(0);
    case 0:
      break;
    case 1:
    case 2:
      ftemp = opt - 1;
      break;
    case 3:
    case 4:
    case 5:
    case 6:
      fbaro = opt - 3;
      break;
    case 7:
    case 8:
      fdepth = opt - 7;
      break;
    case 9:
    case 10:
    case 11:
    case 12:
      fspeed = opt - 9;
      break;
    case 13:
    case 14:			/* bit 7 is flag, bit 3 is value */
      ftd |= 0x80 | ((opt - 13) << 3);
      break;
    case 15:
    case 16:			/* bit 6 is flag, bit 2 is value */
      ftd |= 0x40 | ((opt - 15) << 2);
      break;
    case 'a':
      opt_a = 1;
      break;
    case 'r':
      opt_r = 1;
      break;
    case 'l':
      line = 1;
      break;
    case 't':
      opt_t = 1;
      break;
    case 'e':
      wx200ignoreflags = 1;
      break;
    case 'u':
      wx200ignoreunits = 1;
      break;
    }
  }
  if (line)			/* all vars on a line? */
    for (i = 0; i < 14; i++) {
      current[i] = 1;
    }
  for (i = 0; i < 19; i++) {	/* custom line mode? */
    if (current[i]) line = 1;
  }
  if (optind < argc && argv[optind] != NULL) {
    hostport = argv[optind];
    if ((tmp = strchr(hostport, '%')) != NULL) { /* CGI hack %3A -> : */
      if (*(tmp + 1) == '3' && (*(tmp + 2) == 'A' || *(tmp + 2) == 'a'))
	*tmp = *(tmp + 1) = *(tmp + 2) = ':';
    }
  }

  if ((socket = wx200open(hostport)) < 0) {
    fprintf(stderr, "%s\n", wx200error);
    exit(socket);
  }
  if (query) {
    printf("Content-type: text/html\n");
    if (opt_r)
      printf("Refresh: 10; URL=%s?%s\n",
	     (tmp = getenv("SCRIPT_NAME")) ? tmp : "", query);
    opt_r = 0;
    printf("Status: 200

<html><head>
<title>%s @ %s:%d</title>
</head><body>
<h1>%s @ %s:%d</h1>
<hr>
<pre>
", progname, wx200host, wx200port, progname, wx200host, wx200port);
  }

  wx200bufinit();		/* get current conditions first */
  while ((done < 2) && ((num = wx200bufread(socket)) != EOF)) {
    wx200tab(num);		/* calls wx200parse(num) */
    if ((num == GROUP8) || (num == GROUPF)) done++;
  }

  if (ftemp == -1) ftemp = wx.temp.format; /* prefs override wx200 settings */
  if (fbaro == -1) fbaro = wx.baro.format;
  if (fdepth == -1) fdepth = wx.rain.format;
  if (fspeed == -1) fspeed = wx.wind.format;
  if (ftd & 0x80) ftd |= (ftd & 0x08) >> 2; /* m/d preference? */
  else ftd |= wx.td.format & 0x02;
  if (ftd & 0x40) ftd |= (ftd & 0x04) >> 2; /* 12/24 preference? */
  else ftd |= wx.td.format & 0x01;

  do {				/* forever if -r or --refresh, once otherwise */
    wx200tab(num);		/* calls wx200parse(num) */

    format_data(F_TIME, buf, &wx.td.clock, ftd); /* hack to add seconds */
    strcpy(buf + 20, buf + 11);			/* save am/pm to temp area */
    sprintf(buf + 11, ":%02d", wx.td.clock.sec); /* overwrite with seconds */
    strcpy(buf + 14, buf + 20);			/* append am/pm */
    sprintf(buf + 100, "%2d:%02d%s", ftd & 0x01 ? wx.td.a.hi
	    : (wx.td.a.hi % 12 ? wx.td.a.hi % 12 : 12), wx.td.a.lo,
	    ftd & 0x01 ? "  " : (wx.td.a.hi >= 12 ? "pm" : "am"));

    if (opt_t) {		/* tab-delimited like log mode */
      printf("%s", tabbuf);
      fflush(stdout);
      continue;			/* skip the rest of the loop */
    }

    /* APRS mode contributed by Steve Fraser <sfraser@sierra.apana.org.au> */
    if (current[0] == 2) {
      if (unit_temp(wx.temp.out.n,1) < 0) { /* only put sign if negative */
	sprintf(tmpbuf,"%+02.0f",unit_temp(wx.temp.out.n,1));
      } else {
	sprintf(tmpbuf,"%03.0f",unit_temp(wx.temp.out.n,1));
      }
      printf("%03.0f/%03.0fg%03.0ft%sr%03.0fp%03.0fh%02ib%05.0fXRSW\n",
	     wx.wind.adir.n,
	     unit_speed(wx.wind.aspeed.n,0),
	     unit_speed(wx.wind.gspeed.n,0),
	     tmpbuf,
	     unit_depth(wx.rain.rate.n,1) * 100,
	     unit_depth(wx.rain.yest.n,1) * 100,
	     (int) wx.humid.out.n % 100,
	     wx.baro.local.n * 10);
      fflush(stdout);
      continue;			/* skip the rest of the loop */
    }

    if (line) {			/* custom line mode */
      for (i = 0; i < 19; i++) {
	if (current[i]) {
	  switch (i) {
	  case 0:
	    printf("%s\t", format_data(F_TEMP, buf, &wx.temp.in, ftemp));
	    break;
	  case 1:
	    printf("%s\t", format_data(F_TEMP, buf, &wx.temp.out, ftemp));
	    break;
	  case 2:
	    printf("%s\t", format_data(F_HUMID, buf, &wx.humid.in, 0));
	    break;
	  case 3:
	    printf("%s\t", format_data(F_HUMID, buf, &wx.humid.out, 0));
	    break;
	  case 4:
	    printf("%s\t", format_data(F_TEMP1, buf, &wx.dew.in, ftemp));
	    break;
	  case 5:
	    printf("%s\t", format_data(F_TEMP1, buf, &wx.dew.out, ftemp));
	    break;
	  case 6:
	    printf("%s\t", format_data(F_PRESS, buf, &wx.baro.local, fbaro));
	    break;
	  case 7:
	    printf("%s\t", format_data(F_PRESS, buf, &wx.baro.sea, fbaro));
	    break;
	  case 8:
	    printf("%s\t%s\t",
		   format_data(F_SPEED, buf, &wx.wind.gspeed, fspeed),
		   format_data(F_DIR, buf + 20, &wx.wind.gdir, 0));
	    break;
	  case 9:
	    printf("%s\t%s\t",
		   format_data(F_SPEED, buf, &wx.wind.aspeed, fspeed),
		   format_data(F_DIR, buf + 20, &wx.wind.adir, 0));
	    break;
	  case 10:
	    printf("%s\t", format_data(F_TEMP1, buf, &wx.chill.temp, ftemp));
	    break;
	  case 11:
	    printf("%s\t", format_data(F_RATE, buf, &wx.rain.rate, fdepth));
	    break;
	  case 12:
	    printf("%s\t", format_data(F_DEPTH, buf, &wx.rain.yest, fdepth));
	    break;
	  case 13:
	    printf("%s\t", format_data(F_DEPTH, buf, &wx.rain.total, fdepth));
	    break;
	  case 14:
	    printf("%s\t", trend[wx.baro.trend - 1]);
	    break;
	  case 15:
	    printf("%s\t", pred[wx.baro.pred - 1]);
	    break;
	  case 16:
	    printf("%s\t", acdc[wx.gen.power]);
	    break;
	  case 17:
	    printf("%s\t", batt[wx.gen.lowbat]);
	    break;
	  case 18:
	    printf("%8s/%d/%d\t",
		   disp[wx.gen.section], wx.gen.screen, wx.gen.subscreen);
	    break;
	  }
	}
      }
      printf("\n");
      fflush(stdout);
      continue;			/* skip the rest of the loop */
    }

    printf("WX200 %xf: %20s:%-5d %s", num, wx200host, wx200port, buf);
    if (opt_a)
      printf("    Alarm: %s %s", buf + 100, offon[wx.td.a.on]);
    printf("\n%5s             Display: %8s/%d/%d"
	   "          Power: %s  Battery: %s\n"
	   "             "
	   "---------- Outdoor -----------  ----------- Indoor -----------\n",
	   VERSION, disp[wx.gen.section], wx.gen.screen, wx.gen.subscreen,
	   acdc[wx.gen.power], batt[wx.gen.lowbat]);

    printf(format, "Temperature",
	   format_data(F_TEMP, buf + 60, &wx.temp.out, ftemp),
	   format_data(F_TEMP, buf + 80, &wx.temp.outhi, ftemp),
	   format_data(F_TIME, buf + 100, &wx.temp.outhi, ftd),
	   format_data(F_TEMP, buf, &wx.temp.in, ftemp),
	   format_data(F_TEMP, buf + 20, &wx.temp.inhi, ftemp),
	   format_data(F_TIME, buf + 40, &wx.temp.inhi, ftd),
	   format_data(F_TEMP, buf + 160, &wx.temp.outlo, ftemp),
	   format_data(F_TIME, buf + 180, &wx.temp.outlo, ftd),
	   format_data(F_TEMP, buf + 120, &wx.temp.inlo, ftemp),
	   format_data(F_TIME, buf + 140, &wx.temp.inlo, ftd));
    if (opt_a)
      printf(aformat,
	     format_val(F_TEMP1, buf + 240, wx.temp.outa.nlo, ftemp),
	     format_val(F_TEMP1, buf + 260, wx.temp.outa.nhi, ftemp),
	     offon[wx.temp.outa.on],
	     format_val(F_TEMP1, buf + 200, wx.temp.ina.nlo, ftemp),
	     format_val(F_TEMP1, buf + 220, wx.temp.ina.nhi, ftemp),
	     offon[wx.temp.ina.on]);

    printf(format, "Humidity",
	   format_data(F_HUMID, buf + 60, &wx.humid.out, 0),
	   format_data(F_HUMID, buf + 80, &wx.humid.outhi, 0),
	   format_data(F_TIME, buf + 100, &wx.humid.outhi, ftd),
	   format_data(F_HUMID, buf, &wx.humid.in, 0),
	   format_data(F_HUMID, buf + 20, &wx.humid.inhi, 0),
	   format_data(F_TIME, buf + 40, &wx.humid.inhi, ftd),
	   format_data(F_HUMID, buf + 160, &wx.humid.outlo, 0),
	   format_data(F_TIME, buf + 180, &wx.humid.outlo, ftd),
	   format_data(F_HUMID, buf + 120, &wx.humid.inlo, 0),
	   format_data(F_TIME, buf + 140, &wx.humid.inlo, ftd));
    if (opt_a)
      printf(aformat,
	     format_val(F_HUMID, buf + 240, wx.humid.outa.nlo, 0),
	     format_val(F_HUMID, buf + 260, wx.humid.outa.nhi, 0),
	     offon[wx.humid.outa.on],
	     format_val(F_HUMID, buf + 200, wx.humid.ina.nlo, 0),
	     format_val(F_HUMID, buf + 220, wx.humid.ina.nhi, 0),
	     offon[wx.humid.ina.on]);

    printf(format, "Dew Point",
	   format_data(F_TEMP1, buf + 60, &wx.dew.out, ftemp),
	   format_data(F_TEMP1, buf + 80, &wx.dew.outhi, ftemp),
	   format_data(F_TIME, buf + 100, &wx.dew.outhi, ftd),
	   format_data(F_TEMP1, buf, &wx.dew.in, ftemp),
	   format_data(F_TEMP1, buf + 20, &wx.dew.inhi, ftemp),
	   format_data(F_TIME, buf + 40, &wx.dew.inhi, ftd),
	   format_data(F_TEMP1, buf + 160, &wx.dew.outlo, ftemp),
	   format_data(F_TIME, buf + 180, &wx.dew.outlo, ftd),
	   format_data(F_TEMP1, buf + 120, &wx.dew.inlo, ftemp),
	   format_data(F_TIME, buf + 140, &wx.dew.inlo, ftd));
    if (opt_a)
      printf(aformat,
	     format_val(F_DIFF1, buf + 220, wx.dew.outa.nlo, ftemp), "spread ",
	     offon[wx.dew.outa.on],
	     format_val(F_DIFF1, buf + 200, wx.dew.ina.nlo, ftemp), "spread ",
	     offon[wx.dew.ina.on]);

    printf("Wind Gust:%-6s@%-11s  Wind Gust Hi:%-6s@%-11s %s\n"
	   "      Avg:%-6s@%-11s    Wind Chill:%-6s   Lo:%-6s %s\n",
	   format_data(F_DIR, buf, &wx.wind.gdir, 0),
	   format_data(F_SPEED, buf + 20, &wx.wind.gspeed, fspeed),
	   format_data(F_DIR, buf + 40, &wx.wind.dirhi, 0),
	   format_data(F_SPEED, buf + 60, &wx.wind.speedhi, fspeed),
	   format_data(F_TIME, buf + 80, &wx.wind.speedhi, ftd),
	   format_data(F_DIR, buf + 100, &wx.wind.adir, 0),
	   format_data(F_SPEED, buf + 120, &wx.wind.aspeed, fspeed),
	   format_data(F_TEMP1, buf + 140, &wx.chill.temp, ftemp),
	   format_data(F_TEMP1, buf + 160, &wx.chill.low, ftemp),
	   format_data(F_TIME, buf + 180, &wx.chill.low, ftd));
    if (opt_a)
      printf("    Alarm:%-11s %s            Alarm:%-6s %s\n\n",
	     format_val(F_SPEED, buf + 200,  wx.wind.a.nhi, fspeed),
	     offon[wx.wind.a.on],
	     format_val(F_TEMP1, buf + 220,  wx.chill.a.nlo, ftemp),
	     offon[wx.chill.a.on]);

    printf("Rain Rate:%s  Yesterday:%s Total:%s since %s\n",
	   format_data(F_RATE, buf, &wx.rain.rate, fdepth),
	   format_data(F_DEPTH, buf + 20, &wx.rain.yest, fdepth),
	   format_data(F_DEPTH, buf + 40, &wx.rain.total, fdepth),
	   format_data(F_TIME, buf + 60, &wx.rain.total, ftd));
    if (opt_a)
      printf("    Alarm:%s %s\n\n",
	     format_val(F_RATE, buf + 80, wx.rain.a.nhi, fdepth),
	     offon[wx.rain.a.on]);

    printf("Barometer: %s at %s %s sea; 12-24hr forecast: %s\n",
	   trend[wx.baro.trend - 1],
	   format_data(F_PRESS, buf, &wx.baro.local, fbaro),
	   format_data(F_PRESS, buf + 20, &wx.baro.sea, fbaro),
	   pred[wx.baro.pred - 1]);
    if (opt_a) {
      printf("    Alarm:%s %s",
	     format_val(F_PRESS, buf + 40, wx.baro.a.nhi, fbaro),
	     offon[wx.rain.a.on]);
      if (!wx200ignoreunits)
	printf("      *=Out of Range, !=Sensor Error, #=Both * and !");
      printf("\n");
      fflush(stdout);
    }

  } while (opt_r && (num = wx200bufread(socket)) != EOF);

  if (query)
    printf("</pre>
<hr>
<address>
Weather by <a href=\"http://wx200d.sourceforge.net/\">wx200d</a>
version <b>%s</b>.
</address>
</body></html>
", VERSION);

  exit(wx200close(socket));
}
